tests/component/test_stream_readers.py

```diff
@@ -545,3 +547,591 @@ def test___power_binary_reader___read_many_sample_with_wrong_dtype___raises_erro
+def di_multi_channel_port_uint32_task(
+    task: nidaqmx.Task, sim_6363_device: nidaqmx.system.Device
+) -> nidaqmx.Task:
+    # 6363 port 0 has 32 lines
+    task.di_channels.add_di_chan(
+        flatten_channel_string(sim_6363_device.di_lines[:32].name),
+        line_grouping=LineGrouping.CHAN_PER_LINE,
+    )
+    return task
+
+
+def _get_expected_digital_data_for_sample(num_lines: int, sample_number: int) -> int:
+    # Simulated digital signals "count" from 0 in binary within each group of 8 lines.
+    result = sample_number
+    for _ in range(num_lines // 8):
+        result = (result << 8) | sample_number
+
+    line_mask = (2**num_lines) - 1
+    return result & line_mask
+
+
+def _get_expected_digital_data(num_lines: int, num_samples: int) -> List[int]:
+    return [
+        _get_expected_digital_data_for_sample(num_lines, sample_number)
+        for sample_number in range(num_samples)
+    ]
+
+
+def _get_expected_digital_port_data_for_sample(num_lines: int, sample_number: int) -> List[int]:
+    expected_data = _get_expected_digital_data_for_sample(num_lines, sample_number)
+    # port data is strange, you get full port data masked for only the line in the channel
+    return [expected_data & (1 << line) for line in range(num_lines)]
+
+
+def _get_expected_digital_port_data(num_lines: int, num_samples: int) -> List[List[int]]:
+    result = []
+    # again, port data is strange. it is line-major with the same odd masking/port format as above
+    for line in range(num_lines):
+        result.append(
+            [datum & (1 << line) for datum in _get_expected_digital_data(num_lines, num_samples)]
+        )
+    return result
+
+
+def _bool_array_to_int(bool_array: numpy.ndarray) -> int:
+    result = 0
+    # Simulated data is little-endian
+    for bit in reversed(bool_array):
+        result = (result << 1) | bit
+    return result
+
+
+def test___digital_single_channel_reader___read_one_sample_one_line___returns_valid_samples(
+    di_single_line_task: nidaqmx.Task,
+) -> None:
+    reader = nidaqmx.stream_readers.DigitalSingleChannelReader(di_single_line_task.in_stream)
+    num_lines = di_single_line_task.channels.di_num_lines
+    samples_to_read = 10
+    data = [reader.read_one_sample_one_line() for _ in range(samples_to_read)]
+
+    assert data == _get_expected_digital_data(num_lines, samples_to_read)
+
+
+def test___digital_multi_channel_reader___read_one_sample_one_line___returns_valid_samples(
+    di_single_line_task: nidaqmx.Task,
+) -> None:
+    reader = nidaqmx.stream_readers.DigitalMultiChannelReader(di_single_line_task.in_stream)
+    num_lines = di_single_line_task.channels.di_num_lines
+    samples_to_read = 10
+    data = numpy.full(num_lines, False, dtype=numpy.bool_)
+
+    for sample_number in range(samples_to_read):
+        reader.read_one_sample_one_line(data)
+        assert _bool_array_to_int(data) == _get_expected_digital_data_for_sample(
+            num_lines, sample_number
+        )
```

Comment:
arrange, act, assert

I like how some of these tests cases accumulate data in "act" and validate it in "assert". Can you do the same thing with test cases that read into a preallocated array? This avoids mixing calls to the SUT with validation code, and also when it fails it prints everything, not just the first sample that didn't match.

```
def read_and_copy(read_func, array):
    read_func(array)
    return array.copy()
```

```
sample = numpy.full(num_lines, False, dtype=numpy.bool_)

data = [read_and_copy(reader.read_one_sample_one_line, sample) for _ in range(samples_to_read))]

assert [_bool_array_to_int(sample) for sample in data] == [_get_expected_digital_data_for_sample(num_lines, sample_number) for sample_number in range(samples_to_read)]
```

The assert could also benefit from more helper functions.